### PR TITLE
Back reference's for Task/Application/etc.

### LINKED ISSUE #835: Support backlinks to Other document
Support backlinks to Other document

### PR DESCRIPTION
**Pull request by haiodo**
Configurable Back reference

Screenshot 2022-01-20 at 15 44 36

Screenshot 2022-01-20 at 15 44 50

Closes #835

Signed-off-by: Andrey Sobolev [haiodo@gmail.com](mailto:haiodo@gmail.com)

### CODE REVIEW

File: models/task/src/index.ts [hunk 1]
```diff
@@ -14,53 +14,38 @@
...
-  KanbanTemplateSpace,
-  StateTemplate,
-  DoneStateTemplate,
-  WonStateTemplate,
-  LostStateTemplate,
-  KanbanTemplate,
-  Task,
-  TodoItem
+  DoneState, DoneStateTemplate, Issue, Kanban,
+  KanbanCard, KanbanTemplate, KanbanTemplateSpace, LostState, LostStateTemplate, Project, Sequence, State, StateTemplate, Task,
```
Comment (BykhovDenis) [inline models/task/src/index.ts:40]: I think we need format this...

File: server/core/src/storage.ts [hunk 2]
```diff
@@ -185,7 +185,7 @@ class TServerStorage implements ServerStorage {
   ): Promise<FindResult<T>> {
     return await ctx.with('find-all', {}, (ctx) => {
       const domain = this.hierarchy.getDomain(clazz)
-      if (query.$search !== undefined && query.$search.length > 0) {
+      if (query.$search !== undefined) {
```
Comment (BykhovDenis) [inline server/core/src/storage.ts:188]: Why remove?

File: plugins/task-resources/src/index.ts [hunk 3]
```diff
@@ -93,6 +91,39 @@ async function UnarchiveSpace (object: SpaceWithStates): Promise<void> {
...
+  const named = new Map((await client.findAll(_class, { name: { $like: `%${search}%` } }, { limit: 200 })).map(e => [e._id, e]))
+  const nids: number[] = []
+  if (sequence > 0) {
+    for (let n = 0; n < sequence; n++) {
+      const v = `${n}`
+      if (v.includes(search)) {
+        nids.push(n)
+      }
+    }
+    const numbered = await client.findAll<Task>(_class, { number: { $in: nids } }, { limit: 200 }) as D[]
```
Comment (BykhovDenis) [inline plugins/task-resources/src/index.ts:111]: Is it faster than `{ number: { $like: `%${search}%` } }`?

File: plugins/task-resources/src/index.ts [hunk 4]
```diff
@@ -93,6 +91,39 @@ async function UnarchiveSpace (object: SpaceWithStates): Promise<void> {
...
+
+  // Check number pattern
+
+  const sequence = (await client.findOne(task.class.Sequence, { attachedTo: _class }))?.sequence ?? 0
+
+  const named = new Map((await client.findAll(_class, { name: { $like: `%${search}%` } }, { limit: 200 })).map(e => [e._id, e]))
+  const nids: number[] = []
+  if (sequence > 0) {
+    for (let n = 0; n < sequence; n++) {
+      const v = `${n}`
```
Comment (BykhovDenis) [inline plugins/task-resources/src/index.ts:106]: Why not just `.ToString()` ?

**Review by BykhovDenis — APPROVED** [review]

File: server/core/src/storage.ts (hunk 2 above)
Comment (haiodo) [reply server/core/src/storage.ts:188]: It will fail in mongo if specified.

File: models/task/src/index.ts (hunk 1 above)
Comment (haiodo) [reply models/task/src/index.ts:40]: I think we need to reformat all and add format check instead in separate PR.

**Review by aplatoff — APPROVED** [review]

**Pull request merged**